Cesium3DTilesSelection/src/LayerJsonTerrainLoader.cpp

```diff
@@ -0,0 +1,725 @@
+
+  const QuadtreeTileID* pQuadtreeTileID =
+      std::get_if<QuadtreeTileID>(&tile.getTileID());
+  if (!pQuadtreeTileID) {
+    // This loader only handles QuadtreeTileIDs.
+    return asyncSystem.createResolvedFuture<TileLoadResult>(TileLoadResult{
+        TileUnknownContent{},
+        TileLoadResultState::Failed,
+        0,
+        {}});
+  }
+
+  // Always request the tile from the first layer in which this tile ID is
+  // available.
+  auto firstAvailableIt = this->_layers.begin();
+  while (firstAvailableIt != this->_layers.end() &&
+         !firstAvailableIt->availability.isTileAvailable(*pQuadtreeTileID)) {
+    ++firstAvailableIt;
+  }
+
+  if (firstAvailableIt == this->_layers.end()) {
+    // No layer has this tile available.
+    return asyncSystem.createResolvedFuture<TileLoadResult>(TileLoadResult{
+        TileUnknownContent{},
+        TileLoadResultState::Failed,
+        0,
+        {}});
+  }
+
+  // Start the actual content request.
+  Future<TileLoadResult> futureContent = requestTileContent(
+      pLogger,
+      asyncSystem,
+      pAssetAccessor,
+      *pQuadtreeTileID,
+      tile.getBoundingVolume(),
+      *firstAvailableIt,
+      requestHeaders,
+      contentOptions.enableWaterMask);
+
+  // Also load the same tile in any underlying layers for which this tile
+  // is an availability level. This is necessary because, when we later
+  // create this tile's children, we need to be able to create children
+  // that are only available from an underlying layer, and we can only do
+  // that if we know they're available.
+  std::vector<Future<int>> availabilityRequests;
+
+  auto it = firstAvailableIt;
+  ++it;
+
+  while (it != this->_layers.end()) {
+    if (it->availabilityLevels >= 1 &&
+        (int32_t(pQuadtreeTileID->level) % it->availabilityLevels) == 0 &&
+        it->availability.isTileAvailable(*pQuadtreeTileID)) {
+      availabilityRequests.emplace_back(loadTileAvailability(
+          pLogger,
+          asyncSystem,
+          pAssetAccessor,
+          *pQuadtreeTileID,
+          *it,
+          requestHeaders));
+    }
+  }
+
+  Future<TileLoadResult> futureResult =
+      availabilityRequests.empty()
+          ? std::move(futureContent)
+          : asyncSystem.all(std::move(availabilityRequests))
+                .thenImmediately([futureContent = std::move(futureContent)](
+                                     std::vector<int>&&) mutable {
+                  return std::move(futureContent);
+                });
+
+  return std::move(futureResult)
+      .thenInMainThread(
+          [this, &tile, pQuadtreeTileID](TileLoadResult&& result) {
+            // Now that all our availability is sorted out, create this tile's
+            // children.
+            // TODO: in most cases, we can create this tile's children before
+            // this tile is loaded, and it would be much more efficient to take
+            // advantage of that when we can.
```

Comment:
for most cases, do you mean the cases where the tile is in the middle of the subtree? Because if a tile is at the availability level or at the end of a subtree, it's still need to wait for the current layer to be updated with the new available rectangle before it can create the children right?

Reply:
Yep, exactly right. If availabilityLevels is 10, then there is availability data in tiles at level 0, 10, 20, and 30. At levels 1-9, we don't need to wait for any loading (beyond the level 0 tile) in order to create and traverse those tiles.